src/main/java/com/joejoe2/chat/exception/ChannelDoesNotExist.java

```diff
@@ -1,7 +1,11 @@
 package com.joejoe2.chat.exception;
 
 public class ChannelDoesNotExist extends Exception {
```

Comment:
I think this exception should be a RuntimeException being called `ChannelDoesNotExistException`

Reply:
I do not think exceptions like DoseNotExist should be runtime exception, I have seen other frameworks like Django make these types to be explicitly catched.

Reply:
Agree from the platform the programming language to the programming language perspective might change a little :)
But not drastically though.

Django is a Python framework. Even though in Pythin we have checked and unchecked exceptions checked exceptions are not the same thing as in Java. But Python is not as strict as Java which is a strongly typed programming language with very strict and explicit rules including Exception handling convention.

In Java, we have two types of Exceptions:
* checked - Exception
* unchecked - RuntimtException

Checked exceptions are being checked during the compile time. But also it is impacting how we should operate with these exceptions within the runtime, meaning we need to catch or re-throw these exceptions.

For instance, if we have some internal framework (and now I remember your example - Djanko :) ) and we have some business logic inside and we agree that the client/user us the framework will take care of exception handling that we need to use checked exception meanwhile defining `throws` in the method's signature. In that case, it allows us to delegate this responsibility to the client rather than taking care of exceptions ourselves.
A lot of frameworks in Java follow this approach exactly because of this.
Since most of the time, framework clients/users need to take care of the exception handling and decide how to handle these exceptions.
Just take a look at the JDKs DriverManager#getConnection method which explicitly throws an SQLException which is a checked exception.

At the same time when software is end-user software most of the exceptions are RuntimeExceptions by semantics. Of course, internally we still can design some parts of the API and use checked exceptions, it depends on the design. Although, the use case would be similar to what was mentioned a little above. Quite rare though.

What concerns this particular example then it needs to be a RuntimeException since Java Stream API takes care of throwing this exception.
Although the main thing is not a Stream API but rather semantics and what happens with this exception afterwards.
By semantics and use case - this is a RuntimeExcepton by any means just like most of the exceptions within the end-user application.

Reply:
But the service is used by the controller, making these exceptions checked can force us to handle different cases. For example:
```
    try {
      channelService.block(
          AuthUtil.currentUserDetail().getId(), request.getChannelId(), request.getIsBlocked());
      return ResponseEntity.ok().build();
    } catch (ChannelDoesNotExist e) {
      return new ResponseEntity<>(new ErrorMessageResponse(e.getMessage()), HttpStatus.NOT_FOUND);
    }
```

Reply:
We can move this rather into the service by making the service layer return the ResponseEntity - what a little violates semantics and blurs the boundary between the service and view layer.

The second approach I think would work here the best - we can introduce controller advice here. Also, I'd suggest removing e.getMessage() from the exception to hide internals as much as possible from the HTTP response.

Reply:
The first approach seems unacceptable to me.

The second approach (controller advice) will lower the flexibilities of the service:
1. Consistency Across Contexts: Checked exceptions ensure a consistent error-handling mechanism across various parts of the application, including not just the controller but also command-line interfaces (CLI), task scheduler, or other potential use cases. That's a real experience from my past.
2. Compile-Time Verification: Checked exceptions are verified at compile time, developers never have to dig into the implementation to see the possible unchecked exceptions and add to their controller advice. This is crucial when the service is reused in various ways, as it helps prevent unexpected runtime errors.
3. Flexibility: Applying error-handling mechanisms for these kinds of exceptions in controller advice cannot handle different scenarios for different APIs.

Reply:
Regarding the 1st point - Does this chat service assume CLI to be used?

Definitely, the 2nd point takes place as well - however, as mentioned above in the remark - it depends on how an APOI is being designed. These exceptions are RuntimeExcetion from the semantics even though done like checked ones for now.

3rd point - depends on the use case again. We can't be black and white here. It's grey. Sometimes valid, sometimes not. Also depends on the way this application grows over time. I also feel that having that in the service later would benefit us since we might handle these exceptions differently.

What remains unchanged is that as per semantics and how these exceptions are used and handled - these need to be RuntimeException. We have no issues to handle these ones.

Reply:
`I created these exceptions because I want to notify caller "your action failed due to some reason". And let the caller decide what to do next.` - now I see your reasoning. Yes, these could be checked exceptions.